Magento2/Sniffs/Legacy/ClassReferencesInConfigurationFilesSniff.php

```diff
@@ -0,0 +1,202 @@
+                $phpcsFile->addError(
+                    self::ERROR_MESSAGE_MODULE,
+                    $element->lineNumber - 1,
+                    self::ERROR_CODE_MODULE,
+                );
+            }
+        }
+    }
+
+    /**
+     * Format the incoming XML to avoid tags split into several lines.
+     *
+     * @param File $phpcsFile
+     * @return false|string
+     */
+    private function getFormattedXML(File $phpcsFile)
+    {
+        $doc = new DomDocument('1.0');
+        $doc->formatOutput = true;
+        $doc->loadXML($phpcsFile->getTokensAsString(0, 999999));
+        return $doc->saveXML();
+    }
+
+    /**
+     * Parse an XML for references to PHP class names in selected tags or attributes
+     *
+     * @param SimpleXMLElement $xml
+     * @return array
+     */
+    private function collectClassesInConfig(SimpleXMLElement $xml): array
+    {
+        $classes = $this->getValuesFromXml(
+            $xml,
+            '
+            /config//resource_adapter | /config/*[not(name()="sections")]//class[not(ancestor::observers)]
+                | //model[not(parent::connection)] | //backend_model | //source_model | //price_model
+                | //model_token | //writer_model | //clone_model | //frontend_model | //working_model
+                | //admin_renderer | //renderer',
+            self::FROM_CONTENT
+        );
+        $classes = array_merge(
+            $classes,
+            $this->getValuesFromXml(
+                $xml,
+                '//@backend_model',
+                self::FROM_ATTRIBUTE,
+                'backend_model'
+            )
+        );
+        $classes = array_merge(
+            $classes,
+            $this->getValuesFromXml(
+                $xml,
+                '/config//preference',
+                self::FROM_ATTRIBUTE,
+                'type'
+            )
+        );
+        $classes = array_merge(
+            $classes,
+            $this->getValuesFromXml(
+                $xml,
+                '/logging/*/expected_models/* | /logging/*/actions/*/expected_models/*',
+                self::FROM_NAME
+            )
+        );
+
+        $classes = array_map(
+            function (ExtendedNode $extendedNode) {
+                $extendedNode->value = explode('::', trim($extendedNode->value))[0];
+                return $extendedNode;
+            },
+            $classes
+        );
+
+        return $classes;
+    }
+
+    /**
+     * Extract value from the specified $extractFrom which exist in the XML path
+     *
+     * @param SimpleXMLElement $xml
+     * @param string $xPath
+     * @param int $extractFrom
+     * @param string $attr
+     * @return array
+     */
+    private function getValuesFromXml(SimpleXMLElement $xml, string $xPath, int $extractFrom, string $attr = ''): array
+    {
+        $nodes = $xml->xpath($xPath) ?: [];
+        return array_map(function ($item) use ($extractFrom, $attr) {
+            switch ($extractFrom) {
```

Comment:
Please use 3 private functions instead of the generic function with the switch. The constants are not needed in  this case as well